Bluetooth: Add configuration of CTE in connected mode

ppryga-nordic opened the pull request:
The PR provides set of commits that give possibility to:
- set CTE TX parameters by host,
- extend `pdu_data` structure to store CP bit (CTE present) in S0 byte and CTEInfo in S1 byte of a BT LE data channel packet,
- configure Radio peripheral to transmit CTE

ppryga-nordic committed:
Bluetooth: controller: complete DF TX params set in conn mode

HCI_LE_Set_Connection_CTE_Transmit_Parameters host command was
not completely implemented in controller. There were missign
storage of TX parameters in ll_conn instance.

The commit adds missing part of the command handling.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: radio: add setting of S1 byte in radio packet

To transmit CTE with data channel packet the CTE configuration must
be stored in S1 field, that is placed just after length field.

The commit changes radio_pkt_configure function to allow configuration
of S1 field. When S1 field is used to store information about CTE
then its length is 8 bits. The content of the field is filled with
CTEInfo data. The same as the one used in periodic advertising.

The signature of the radio_pkt_configure function was not changed.
There were not used bits in flags parameter of the function.
From now on, bit 3 of the flags parameter will be used to store
information whether CTE will be added to the packet.

Besides that changed, the commit provides set of macros that
help in preparation and validation of the flags parameter content.
Macros provide information about bits that are currently used.
Ther are also macros that help to retrieve particular information
stored in the flags parameter.

Every place of code where the radio_pkt_configure function occurs
was changed to use provided macros.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: lll: add TX configuration of CTE in conn mode

Add configuration of transmission of CTE in connected mode.
The transmission is enabled only for PDUs that have CTE present (CP)
bit set to 1.
Radio peripheral is configued by df_cte_tx_configure to transmit CTE.
The same parameters are set for connected and connectionless mode.
The function was changed to re-use the same code for configuration
of CTE in both modes.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: radio: add fun to switch to RX after PHYEND evt

When PDU has CTE added to the end, the PHYEND event must be used
to correclty switch between TX and RX. END event is triggered just
before start of CTE.
The commit adds new function that configures radio to disable after
PHYEND is generated and then swtich to RX.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Add CTE present bit to pdu_data

If the CTE is attached to data channel packet, the packet must
include information about CTE. The information is stored in
CTEInfo structure that is available in S1 byte of a packet.
Radio checks if the S1 byte is present by verification of
CTE present (CP) bit in byte S0.

The commit add these data to pdu_data structure declaration.
That allows to prepare packet before it is parsed and send
by Radio peripheral.
Besides that there added a new function ull_pdu_data_init that
initlializes fields in pdu_data object that are read only by
lower link layer. CP bit in S0 byte and content of S1 bytes
are examples of such fields.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -288,7 +288,8 @@ void radio_aa_set(uint8_t *aa)
 void radio_pkt_configure(uint8_t bits_len, uint8_t max_len, uint8_t flags)
 {
-	uint8_t dc = flags & 0x01; /* Adv or Data channel */
+	uint8_t pdu_type = RADIO_PKT_CONF_PDU_TYPE_GET(flags); /* Adv or Data channel */
```
```suggestion
	const uint8_t pdu_type = RADIO_PKT_CONF_PDU_TYPE_GET(flags); /* Adv or Data channel */
```

ppryga replied:
Corrected

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -334,31 +337,34 @@ void radio_pkt_configure(uint8_t bits_len, uint8_t max_len, uint8_t flags)
 			  RADIO_PCNF0_S1INCL_Pos) & RADIO_PCNF0_S1INCL_Msk;
+#if defined(CONFIG_BT_CTLR_DF)
+		if (RADIO_PKT_CONF_CTE_GET(flags) == RADIO_PKT_CONF_CTE_ENABLED) {
+			bits_s1 = 8U;
```
What does `_s1` mean in this context? 

And what does the value `8U` mean?

ppryga replied:
`S1` is a name of a byte used in Radii peripheral SOCs documentation. 
`bits` simitlar to `bits_len` means nubmer of bits in the `S1` byte. Hence 8U for 8 bits.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -334,31 +337,34 @@ void radio_pkt_configure(uint8_t bits_len, uint8_t max_len, uint8_t flags)
-			     RADIO_PCNF1_ENDIAN_Msk);
+	NRF_RADIO->PCNF1 |=
+		((((uint32_t)max_len) << RADIO_PCNF1_MAXLEN_Pos) & RADIO_PCNF1_MAXLEN_Msk) |
+		(((0UL) << RADIO_PCNF1_STATLEN_Pos) & RADIO_PCNF1_STATLEN_Msk) |
```
Not sure if these lines were actually changed, or if you just formatted them, but isn't it redundant to do `<<` on a 0 value?

ppryga replied:
It was just re-formatted. I think that it is just for sake of readability, to make the configuration more explicit even though it is redundant. Compiler will optimize it anyway.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.h:
```diff
@@ -5,6 +5,48 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+/* Set of macros related with Radio packet configuration flags */
+#define RADIO_PKT_CONF_
```
Is it normal to add this sort of define for the radio headers?

ppryga replied:
This is a first time, but these macros are used outside of the `radio.c` so I think there is no better place to put them.

Thalley replied:
I was more thinking about the explicit empty macro `RADIO_PKT_CONF_`. What purpose does it serve?

ppryga replied:
ah this is a leftover... sorry I didn't spot that.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.h:
```diff
@@ -5,6 +5,48 @@
+#define RADIO_PKT_CONF_PDU_TYPE_DC (1U)
+/* PHY type, two bit field */
+#define RADIO_PKT_CONF_PHY_POS (1U)
+#define RADIO_PKT_CONF_PHY_MSK (BIT_MASK(2) << RADIO_PKT_CONF_PHY_POS)
```
Should it have been
```suggestion
#define RADIO_PKT_CONF_PHY_MSK (BIT_MASK(2U) << RADIO_PKT_CONF_PHY_POS)
```
Now that all other values are suffixed with `U`?

ppryga replied:
corrected

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -363,6 +363,23 @@ void radio_switch_complete_and_phy_end_b2b_tx(uint8_t phy_curr, uint8_t flags_cu
+	NRF_RADIO->SHORTS = RADIO_SHORTS_READY_START_Msk | RADIO_SHORTS_PHYEND_DISABLE_Msk;
+
+	/* NOTE: As Tx chain delays are negligible constant values (~1 us)
+	 *	 across nRF5x radios, sw_switch assumes the 1M chain delay for
+	 *       calculations.
```
```suggestion
	 *       across nRF5x radios, sw_switch assumes the 1M chain delay for
	 *       calculations.
```

ppryga replied:
Corrected

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:
```diff
@@ -894,10 +894,11 @@ static int prepare_cb(struct lll_prepare_param *p)
 	/* TODO: if coded we use S8? */
 	radio_phy_set(lll->phy_p, lll->phy_flags);
-	radio_pkt_configure(8, PDU_AC_LEG_PAYLOAD_SIZE_MAX, (lll->phy_p << 1));
+	radio_pkt_configure(8, PDU_AC_LEG_PAYLOAD_SIZE_MAX, RADIO_PKT_CONF_PHY(lll->phy_p));
```
Should this also have been
```suggestion
	radio_pkt_configure(RADIO_PKT_CONF_LENGTH_8BIT, PDU_AC_LEG_PAYLOAD_SIZE_MAX, RADIO_PKT_CONF_PHY(lll->phy_p));
```
?

ppryga replied:
Corrected, thanks for catching this!

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -246,6 +253,17 @@ void lll_conn_isr_rx(void *param)
+	} else
+#endif /* CONFIG_BT_CTRL_DF_CONN_CTE_TX */
+	{
+		cte_len = 0;
```
Should we do this for all unsigned values? 

```suggestion
		cte_len = 0U;
```

ppryga replied:
corrected

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -310,11 +340,11 @@ void lll_conn_isr_rx(void *param)
 				 HAL_RADIO_GPIO_PA_OFFSET);
 #else /* !CONFIG_BT_CTLR_PHY */
-	radio_gpio_pa_lna_enable(radio_tmr_tifs_base_get() + EVENT_IFS_US -
+	radio_gpio_pa_lna_enable(radio_tmr_tifs_base_get() + EVENT_IFS_US + cte_len -
```
`radio_tmr_tifs_base_get() + EVENT_IFS_US + cte_len` and `- HAL_RADIO_GPIO_PA_OFFSET` is common between the two calls to `radio_gpio_pa_lna_enable`. We could consider moving the common calculation to a common variable, and only change the input of `radio_rx_chain_delay_get` for these calls.

ppryga replied:
Changes added

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -427,9 +461,21 @@ void lll_conn_isr_tx(void *param)
 	/* +/- 2us active clock jitter, +1 us hcto compensation */
-	hcto = radio_tmr_tifs_base_get() + EVENT_IFS_US + 4 +
-		RANGE_DELAY_US + 1;
+	hcto = radio_tmr_tifs_base_get() + EVENT_IFS_US + 4 + RANGE_DELAY_US + 1;
```
There seems to be quite a few lines in this PR that has only been reformatted and haven't really changed. I suggest that you avoid doing that, so that your PRs will be more clear and easier to review, as well as given a clearer git history :)

ppryga replied:
There are some lines that were just re-formatted. I didn't catch them, I'll put more attention to that in future.

ppryga replied:
Yes it is. I've updated it to: `(EVENT_CLOCK_JITTER_US << 1)` and the `+ 1` at the end to: `HCTO_START_DELAY_US`

Thalley commented on subsys/bluetooth/controller/ll_sw/pdu.h:
```diff
@@ -516,6 +516,11 @@ enum pdu_data_llctrl_type {
 	PDU_DATA_LLCTRL_TYPE_UNUSED = 0xFF
 };
 
+enum pdu_data_cte_present {
+	PDU_DATA_CTE_PRESENT_BIT_DISABLED = 0x0,
+	PDU_DATA_CTE_PRESENT_BIT_ENABLED = 0x1
+};
```
This looks like it could just be a boolean value instead

ppryga replied:
enum removed, I'll use 'true` or `false` instead.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -973,22 +971,55 @@ static uint8_t cte_info_clear(struct ll_adv_set *adv, struct lll_df_adv_cfg *df_
  */
 uint8_t ll_df_set_conn_cte_tx_params(uint16_t handle, uint8_t cte_types, uint8_t switch_pattern_len,
-				     uint8_t *ant_id)
+				     uint8_t *ant_ids)
```
```suggestion
				     const uint8_t *ant_ids)
```

ppryga replied:
Corrected

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -973,22 +971,55 @@ static uint8_t cte_info_clear(struct ll_adv_set *adv, struct lll_df_adv_cfg *df_
+		return BT_HCI_ERR_UNSUPP_FEATURE_PARAM_VAL;
+	}
+
+	memcpy(df_tx_cfg->ant_ids, ant_ids, switch_pattern_len);
```
```suggestion
	(void)memcpy(df_tx_cfg->ant_ids, ant_ids, switch_pattern_len);
```

ppryga replied:
Corrected

ppryga-nordic committed:
tests: Bluetooth: df: Add tests for set conn CTE TX params

Add unit tests for HCI command HCI_LE_Set_Connection_CTE_-
Transmit_Parameters.

Part of a code that was available in df/connection_cte_req/common.c,
related with artificial handling of connection, is moved to
df/common/bt_conn_common.c.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: df: remove duplicated code from CTE req tests

There is a common implementation of connection handling code
that may be shared by tests: connection_cte_req and connection_cte-
tx_params.
This commit removes code implemented in common.c and common.h
files. The same code is currently available in ../df/common/
directory.
All calls were updated to use function from common implementation.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
Added two new commits:
- UT for HCI_LE_Set_Connection_CTE_Transmit_Parameters
- aligned UT in connection_cte_req to use common code from ../df/common/ directory.

@Thalley I've addressed your comments, thank you for fast review!

Thalley commented on tests/bluetooth/df/connection_cte_tx_params/Kconfig:
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2020 Nordic Semiconductor ASA
```
```suggestion
# Copyright (c) 2021 Nordic Semiconductor ASA
```

ppryga replied:
Corrected

Thalley commented on tests/bluetooth/df/connection_cte_tx_params/prj.conf:
```diff
@@ -0,0 +1,31 @@
+CONFIG_BT=y
+
+CONFIG_BT_CENTRAL=y
+# CONFIG_BT_SMP=y
+# CONFIG_BT_GATT_CLIENT=y
```
Should these and the others that are commented out from this file be removed?

ppryga replied:
Yes they should. I've forgotten to clean it up after debugging.

Thalley commented on tests/bluetooth/df/connection_cte_tx_params/src/test_set_conn_cte_tx_params.c:
```diff
@@ -0,0 +1,205 @@
+
+static uint16_t g_conn_handle;
+
+static uint8_t g_ant_ids[] = { 0x1, 0x2, 0x3, 0x4, 0x5 };
+
+/**
+ * @brief Structure to store CTE receive parameters for unit tests setup.
+ */
+struct ut_bt_df_conn_cte_tx_params {
+	uint8_t cte_types;
+	uint8_t switch_pattern_len;
+	uint8_t *ant_ids;
+};
```
```suggestion
static const uint8_t g_ant_ids[] = { 0x1, 0x2, 0x3, 0x4, 0x5 };

/**
 * @brief Structure to store CTE receive parameters for unit tests setup.
 */
struct ut_bt_df_conn_cte_tx_params {
	uint8_t cte_types;
	uint8_t switch_pattern_len;
	const uint8_t *ant_ids;
};
```

Thalley commented on tests/bluetooth/df/connection_cte_tx_params/src/test_set_conn_cte_tx_params.c:
```diff
@@ -0,0 +1,205 @@
+	cp->handle = sys_cpu_to_le16(conn_handle);
+	cp->cte_types = params->cte_types;
+
+	uint8_t *dest_ant_ids = net_buf_add(buf, params->switch_pattern_len);
```
Consider moving the declaration to top of scope as has been done for other declarations in this file.

ppryga replied:
Done, but I prefer C++ approach, declare as close as possible to place where a variable is used.

Thalley replied:
> Done, but I prefer C++ approach, declare as close as possible to place where a variable is used.

Given that we apparently don't have any rules for that, and that there is some different opinions on it, it was fine to keep as is, but then the `err` declarations should have been moved instead :)

If there are no rules, then at least we can be consistent :D

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -500,7 +548,9 @@ void lll_conn_rx_pkt_set(struct lll_conn *lll)
 #if defined(CONFIG_BT_CTLR_LE_ENC)
 	} else if (lll->enc_rx) {
-		radio_pkt_configure(8, (max_rx_octets + 4), (phy << 1) | 0x01);
+		radio_pkt_configure(RADIO_PKT_CONF_LENGTH_8BIT, (max_rx_octets + 4),
```
This is a magic number for me also. I've just moved existing code. 
@cvinayak what is the origin of the number?

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -973,22 +971,56 @@ static uint8_t cte_info_clear(struct ll_adv_set *adv, struct lll_df_adv_cfg *df_
+	}
+
+	/* Bits other than representing AoA, AoD 1us, AoD 2us are RFU */
+	if (cte_types == 0 ||
```
Corrected, thanks for spotting that!

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -973,22 +971,56 @@ static uint8_t cte_info_clear(struct ll_adv_set *adv, struct lll_df_adv_cfg *df_
+	/* Bits other than representing AoA, AoD 1us, AoD 2us are RFU */
+	if (cte_types == 0 ||
+	    ((cte_types & (~(uint8_t)(BT_HCI_LE_AOA_CTE_RSP | BT_HCI_LE_AOD_CTE_RSP_1US |
+				      BT_HCI_LE_AOD_CTE_RSP_2US))) != 0)) {
```
Corrected

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -309,15 +342,14 @@ void lll_conn_isr_rx(void *param)
-	radio_gpio_pa_lna_enable(radio_tmr_tifs_base_get() + EVENT_IFS_US -
-				 radio_rx_chain_delay_get(lll->phy_rx, 1) -
-				 HAL_RADIO_GPIO_PA_OFFSET);
+	pa_lna_enable_us -= radio_rx_chain_delay_get(lll->phy_rx, 1);
```
Thank , I've forgotten that we already have these macros.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -309,15 +342,14 @@ void lll_conn_isr_rx(void *param)
-	radio_gpio_pa_lna_enable(radio_tmr_tifs_base_get() + EVENT_IFS_US -
-				 radio_rx_chain_delay_get(0, 0) -
-				 HAL_RADIO_GPIO_PA_OFFSET);
+	pa_lna_enable_us -= radio_rx_chain_delay_get(0, 0);
```
Corrected

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -427,9 +463,21 @@ void lll_conn_isr_tx(void *param)
+	if (pdu_tx->cp) {
+		cte_len = CTE_LEN_US(pdu_tx->cte_info.time);
+	} else {
+		cte_len = 0;
```
Corrected

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -547,17 +599,26 @@ void lll_conn_tx_pkt_set(struct lll_conn *lll, struct pdu_data *pdu_data_tx)
 	} else if (lll->enc_tx) {
-		radio_pkt_configure(8, (max_tx_octets + 4U),
-				    (phy << 1) | 0x01);
+		radio_pkt_configure(RADIO_PKT_CONF_LENGTH_8BIT, (max_tx_octets + 4U), pkt_flags);
```
Ok. I've got information from @cvinayak , it is MIC size. I'll use `PDU_MIC_SIZE` instead of `4`

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -864,4 +921,33 @@ static void empty_tx_init(void)
+
+	link = memq_peek(lll->memq_tx.head, lll->memq_tx.tail, (void **)&tx);
+	if (lll->empty || !link) {
+		p = (void *)radio_pkt_empty_get();
```
removed unnecessary cast

ppryga commented:
Fixed compliance issue.

ppryga commented:
Fixed build issues.

carlescufi commented:
@cvinayak can you take another look?

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_central.c:
```diff
@@ -25,6 +26,7 @@
 #include "lll_vendor.h"
 #include "lll_clock.h"
 #include "lll_df_types.h"
+#include "lll_df_internal.h"
```
Move this down with other `_internal.h`

ppryga replied:
Moved

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -27,6 +28,7 @@
 #include "lll.h"
 #include "lll_clock.h"
 #include "lll_df_types.h"
+#include "lll_df_internal.h"
```
Move alongwith other `_internal.h`

ppryga replied:
Moved

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_tim_internal.h:
```diff
@@ -12,6 +12,12 @@
 #define RANGE_DELAY_US (2 * RANGE_DISTANCE * 4 / 1000)
 
+/* This is a compensation of delay between PPI task START and timer start counting.
+ * HACTO is a timer used to stop Radio peripheral in receive mode if packet address was not
```
```suggestion
 * HCTO is a timer used to stop Radio peripheral in receive mode if packet address was not
```

ppryga replied:
Corrected

cvinayak reviewed (changes requested):
Minor changes requested, otherwise looks good to me.

ppryga commented:
@cvinayak requested changes are done, please take another look.
